diff --git a/component/process/process_freebsd_amd64.go b/component/process/process_freebsd_x86.go
similarity index 92%
rename from component/process/process_freebsd_amd64.go
rename to component/process/process_freebsd_x86.go
index cb43c5ff..5e6d923c 100644
--- a/component/process/process_freebsd_amd64.go
+++ b/component/process/process_freebsd_x86.go
@@ -1,3 +1,5 @@
+//go:build freebsd && (386 || amd64)
+
 package process
 
 import (
@@ -177,33 +179,29 @@ func (s *searcher) searchSocketPid(socket uint64) (uint32, error) {
 	return 0, ErrNotFound
 }
 
-func newSearcher(major int) *searcher {
+func newSearcher(major int, machine string) *searcher {
 	var s *searcher
 	switch major {
-	case 11:
-		s = &searcher{
-			headSize:     32,
-			tcpItemSize:  1304,
-			udpItemSize:  632,
-			port:         198,
-			ip:           228,
-			vflag:        116,
-			socket:       88,
-			fileItemSize: 80,
-			data:         56,
-			pid:          8,
-			udpInpOffset: 8,
-		}
 	case 12:
 		fallthrough
 	case 13:
+		fallthrough
+	case 14:
+		fallthrough
+	case 15:
+		var vflag int
+		if machine == "i386" {
+			vflag = 388
+		} else {
+			vflag = 392
+		}
 		s = &searcher{
 			headSize:     64,
 			tcpItemSize:  744,
 			udpItemSize:  400,
 			port:         254,
 			ip:           284,
-			vflag:        392,
+			vflag:        vflag,
 			socket:       16,
 			fileItemSize: 128,
 			data:         56,
@@ -218,6 +216,10 @@ func initSearcher() error {
 	if err != nil {
 		return err
 	}
+	machine, err := syscall.Sysctl("hw.machine")
+	if err != nil {
+		return err
+	}
 
 	dot := strings.Index(osRelease, ".")
 	if dot != -1 {
@@ -227,7 +229,7 @@ func initSearcher() error {
 	if err != nil {
 		return err
 	}
-	defaultSearcher = newSearcher(major)
+	defaultSearcher = newSearcher(major, machine)
 	if defaultSearcher == nil {
 		return fmt.Errorf("unsupported freebsd version %d", major)
 	}
diff --git a/component/process/process_other.go b/component/process/process_other.go
index eea6e5fd..4070c467 100644
--- a/component/process/process_other.go
+++ b/component/process/process_other.go
@@ -1,4 +1,4 @@
-//go:build !darwin && !linux && !windows && (!freebsd || !amd64)
+//go:build !darwin && !linux && !windows && (!freebsd || !(amd64 || 386))
 
 package process
 
diff --git a/config/config.go b/config/config.go
index d79ab65c..4dc0c0ca 100644
--- a/config/config.go
+++ b/config/config.go
@@ -8,6 +8,7 @@ import (
 	"net/netip"
 	"net/url"
 	"path/filepath"
+	"runtime"
 	"strings"
 	"time"
 	_ "unsafe"
@@ -678,6 +679,13 @@ func ParseRawConfig(rawCfg *RawConfig) (*Config, error) {
 	if err != nil {
 		return nil, err
 	}
+	// check dns settings
+	limitDNS := config.General.Tun.Enable && config.General.Tun.AutoRoute && runtime.GOOS == "freebsd"
+	if limitDNS {
+		if dnsCfg.Listen != "127.0.0.1:53" && dnsCfg.Listen != "0.0.0.0:53" {
+			return nil, fmt.Errorf("the option `dns.listen` should be 127.0.0.1:53 or 0.0.0.0:53 when automatic routing is enabled")
+		}
+	}
 
 	err = parseTuicServer(rawCfg.TuicServer, config.General)
 	if err != nil {
diff --git a/listener/config/tun.go b/listener/config/tun.go
index 4e12056d..b3fe56d4 100644
--- a/listener/config/tun.go
+++ b/listener/config/tun.go
@@ -36,6 +36,7 @@ type Tun struct {
 	Inet6Address           []netip.Prefix `yaml:"inet6-address" json:"inet6-address,omitempty"`
 	IPRoute2TableIndex     int            `yaml:"iproute2-table-index" json:"iproute2-table-index,omitempty"`
 	IPRoute2RuleIndex      int            `yaml:"iproute2-rule-index" json:"iproute2-rule-index,omitempty"`
+	FIBIndex               int            `yaml:"fib-index" json:"fib-index,omitempty"`
 	AutoRedirect           bool           `yaml:"auto-redirect" json:"auto-redirect,omitempty"`
 	AutoRedirectInputMark  uint32         `yaml:"auto-redirect-input-mark" json:"auto-redirect-input-mark,omitempty"`
 	AutoRedirectOutputMark uint32         `yaml:"auto-redirect-output-mark" json:"auto-redirect-output-mark,omitempty"`
diff --git a/listener/sing_tun/server.go b/listener/sing_tun/server.go
index 6998f0c1..740dd89f 100644
--- a/listener/sing_tun/server.go
+++ b/listener/sing_tun/server.go
@@ -188,6 +188,10 @@ func New(options LC.Tun, tunnel C.Tunnel, additions ...inbound.Addition) (l *Lis
 	if ruleIndex == 0 {
 		ruleIndex = tun.DefaultIPRoute2RuleIndex
 	}
+	fibIndex := options.FIBIndex
+	if fibIndex == 0 {
+		fibIndex = tun.DefaultFIBIndex
+	}
 	inputMark := options.AutoRedirectInputMark
 	if inputMark == 0 {
 		inputMark = tun.DefaultAutoRedirectInputMark
@@ -345,6 +349,7 @@ func New(options LC.Tun, tunnel C.Tunnel, additions ...inbound.Addition) (l *Lis
 		AutoRoute:                options.AutoRoute,
 		IPRoute2TableIndex:       tableIndex,
 		IPRoute2RuleIndex:        ruleIndex,
+		FIBIndex:                 fibIndex,
 		AutoRedirectInputMark:    inputMark,
 		AutoRedirectOutputMark:   outputMark,
 		StrictRoute:              options.StrictRoute,
