diff --git a/monitor_other.go b/monitor_other.go
index 40a936e..f4e0b33 100644
--- a/monitor_other.go
+++ b/monitor_other.go
@@ -1,4 +1,4 @@
-//go:build !(linux || windows || darwin)
+//go:build !(linux || windows || darwin || freebsd)
 
 package tun
 
diff --git a/monitor_shared.go b/monitor_shared.go
index c2b4254..004700c 100644
--- a/monitor_shared.go
+++ b/monitor_shared.go
@@ -1,4 +1,4 @@
-//go:build linux || windows || darwin
+//go:build linux || windows || darwin || freebsd
 
 package tun
 
diff --git a/tun.go b/tun.go
index fdcc548..4e38fd9 100644
--- a/tun.go
+++ b/tun.go
@@ -68,6 +68,7 @@ type DarwinTUN interface {
 const (
 	DefaultIPRoute2TableIndex = 2022
 	DefaultIPRoute2RuleIndex  = 9000
+	DefaultFIBIndex           = 2022
 )
 
 type Options struct {
@@ -82,6 +83,7 @@ type Options struct {
 	DNSServers               []netip.Addr
 	IPRoute2TableIndex       int
 	IPRoute2RuleIndex        int
+	FIBIndex                 int
 	AutoRedirectMarkMode     bool
 	AutoRedirectInputMark    uint32
 	AutoRedirectOutputMark   uint32
@@ -129,6 +131,8 @@ func (o *Options) Inet4GatewayAddr() netip.Addr {
 				return o.Inet4Address[0].Addr().Next()
 			}
 		case "darwin":
+			fallthrough
+		case "freebsd":
 			return o.Inet4Address[0].Addr()
 		default:
 			if HasNextAddress(o.Inet4Address[0], 1) {
@@ -153,6 +157,8 @@ func (o *Options) Inet6GatewayAddr() netip.Addr {
 				return o.Inet6Address[0].Addr().Next()
 			}
 		case "darwin":
+			fallthrough
+		case "freebsd":
 			return o.Inet6Address[0].Addr()
 		default:
 			if HasNextAddress(o.Inet6Address[0], 1) {
diff --git a/tun_nondarwin.go b/tun_nondarwin.go
index 0faa2c9..053b931 100644
--- a/tun_nondarwin.go
+++ b/tun_nondarwin.go
@@ -1,4 +1,4 @@
-//go:build !darwin
+//go:build !darwin && !freebsd
 
 package tun
 
diff --git a/tun_other.go b/tun_other.go
index 1db48f9..432d26b 100644
--- a/tun_other.go
+++ b/tun_other.go
@@ -1,4 +1,4 @@
-//go:build !(linux || windows || darwin)
+//go:build !(linux || windows || darwin || freebsd)
 
 package tun
 
diff --git a/tun_rules.go b/tun_rules.go
index 7adc948..5487fdb 100644
--- a/tun_rules.go
+++ b/tun_rules.go
@@ -101,7 +101,7 @@ func buildExcludedRanges(includeRanges []ranges.Range[uint32], excludeRanges []r
 	return ranges.Merge(uidRanges)
 }
 
-const autoRouteUseSubRanges = runtime.GOOS == "darwin"
+const autoRouteUseSubRanges = runtime.GOOS == "darwin" || runtime.GOOS == "freebsd"
 
 func (o *Options) BuildAutoRouteRanges(underNetworkExtension bool) ([]netip.Prefix, error) {
 	var routeRanges []netip.Prefix
